jps-plugin/src/org/intellij/erlang/jps/builder/ErlangModuleBuildOrderBuilder.java

```diff
@@ -0,0 +1,199 @@
+    LOG.info("Computing dirty files");
+    LOG.debug("Load project build order.");
+    ErlangProjectBuildOrder projectBuildOrder = loadProjectBuildOrder(context);
+    if (projectBuildOrder == null) {
+      addPrepareDependenciesFailedMessage(context);
+      return;
+    }
+
+    LOG.debug("Collect dirty files.");
+    List<String> dirtyErlangFilePaths = new DirtyFileProcessor<String, ErlangModuleBuildOrderTarget>() {
+      @Nullable
+      @Override
+      protected String getDirtyElement(@NotNull File file) throws IOException {
+        return isSourceOrHeader(file) ? file.getAbsolutePath() : null;
+      }
+    }.collectDirtyElements(holder);
+
+    List<ErlangTarget> erlangTargets = ContainerUtil.mapNotNull(context.getProjectDescriptor().getBuildTargetIndex().getAllTargets(), new Function<BuildTarget<?>, ErlangTarget>() {
+      @Nullable
+      @Override
+      public ErlangTarget fun(BuildTarget<?> buildTarget) {
+        return buildTarget instanceof ErlangTarget ? (ErlangTarget) buildTarget : null;
+      }
+    });
+    for (ErlangTarget erlangTarget : erlangTargets) {
+      erlangTarget.setBuildOrder(new ErlangModuleBuildOrder());
+    }
+
+
+    if (dirtyErlangFilePaths.isEmpty()) {
+      LOG.debug("There aren't dirty .erl or .hrl files.");
+    }
+    else {
+      LOG.debug("Search dirty modules.");
+      List<String> sortedDirtyModules = getSortedDirtyModules(context, projectBuildOrder, dirtyErlangFilePaths);
+      addFilesToBuiltTarget(context, sortedDirtyModules);
+    }
+
+  }
+
+  @NotNull
+  @Override
+  public String getPresentableName() {
+    return NAME;
+  }
+
+  @Nullable
+  private static ErlangProjectBuildOrder loadProjectBuildOrder(@NotNull CompileContext context) {
+    return readFromXML(context, BUILD_ORDER_FILE_NAME, ErlangProjectBuildOrder.class);
+  }
+
+  private static boolean isSourceOrHeader(@NotNull File file) {
+    return file.getName().endsWith(".erl") || file.getName().endsWith(".hrl");
+  }
+
+  private static List<String> getSortedDirtyModules(@NotNull CompileContext context,
+                                                    @NotNull ErlangProjectBuildOrder projectBuildOrder,
+                                                    @NotNull List<String> dirtyErlangFilePaths) {
+    SortedModuleDependencyGraph graph = new SortedModuleDependencyGraph(projectBuildOrder);
+    Set<String> allDirtyFiles = getAllDirtyFiles(dirtyErlangFilePaths, GraphGenerator.create(graph));
+    return getSortedDirtyModules(projectBuildOrder.myErlangFiles, allDirtyFiles);
+  }
+
+
+  private static void addPrepareDependenciesFailedMessage(@NotNull CompileContext context) {
+    context.processMessage(new CompilerMessage(NAME, BuildMessage.Kind.WARNING, "The project will be fully rebuilt due to errors."));
+  }
+
+  private static Set<String> getAllDirtyFiles(@NotNull List<String> dirtyFiles,
+                                              @NotNull GraphGenerator<String> graph) {
```

Comment:
It would make much more sense if this method had an ErlangProjectBuildOrder parameter. Here is why: this guy's responsibility is to build this all dirty files list - regardless of how it will do it.